### ISSUE TITLE
Flax overhead

### ISSUE THREAD

**Issue by GJBoth** [google/flax#1080]
I'm comparing the speed of performing a simple dot product with a fixed matrix 'naked' or in a flax layer. I noticed that by using flax, I'm seeing a massively decreased performance: from approximately 350 microsecond to 5-10ms. Here's a [colab](https://colab.research.google.com/drive/1A6fczIz02zruboM7JpujNjUqao6PrWVx?usp=sharing) where I reproduce the issue. I compare a naked approach, flax (both implicit and explicit) and a python dataclass. Only with the Flax classes do I see the performance hit. 

Am I timing wrong (I used .block_until_ready() everywhere), or am I making a wrong comparison? Or is it something with Flax?

**Comment by jheek** [comment]
Unlike for example PyTorch we haven't put much effort in reducing overhead. This has a few reasons:
1. Python is slow, so fast code eventually tends to move to c++ making the implementation much more complex
2. Jax itself has large dispatch overheads
3. Most importantly, the dispatch overheads disappear  once you compile you code. So using `jax.jit` on any of the examples you gave should lead to the same compiled program being generated with the exact same performance.

The reason why in this case the overhead is so high is because similar to other frameworks the compute is done async from the python thread. Normally the tensors are big enough to mask any overhead from the python interpreter. Still `jax.jit` compiled code can be an order of magnitude faster in many cases.
